### PR TITLE
Add Listener to PreAssetsCompileEvent

### PR DESCRIPTION
**Pull request by WebMamba**

### CODE REVIEW

**Comment by WebMamba** [comment]
The `PublicAssetsPathResolverInterface` had some changes on the last release of AssetMapper. I'm not sure how we should handle it here. See the error on phpstan. Should we just remove the `getPublicFilesystemPath` from the `SassPublicPathAssetPathResolver`? What do you think @weaverryan?

**Comment by weaverryan** [comment]
I see it. Could we surround that with a method_exists? Would that do it?

```php
public function getPublicFilesystemPath(): string
{
    if (!method_exists($this->decorator', 'getPublicFilesystemPath')) {
        throw new \Exception('...');
    }

    // rest of method
}
```

File: src/Listener/PreAssetsCompileEventListener.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            )
+        );
+
+        $process = $this->sassBuilder->runBuild(false);
+
+        if ($process->isSuccessful()) {
+            return;
+        }
+
+        throw new \RuntimeException($process->getErrorOutput());
```
Comment (weaverryan) [inline src/Listener/PreAssetsCompileEventListener.php:38]: Maybe clarify what process failed:

```suggestion
        throw new \RuntimeException(sprintf('Error compiling sass: "%s", $process->getErrorOutput()));
```

**Comment by WebMamba** [comment]
Damm didn't know about this trick to fool phpstan! Thanks! 😁

File: src/AssetMapper/SassPublicPathAssetPathResolver.php [hunk 2]
```diff
@@ -30,6 +30,10 @@ public function resolvePublicPath(string $logicalPath): string
 
     public function getPublicFilesystemPath(): string
     {
+        if (!method_exists($this->decorator, 'getPublicFilesystemPath')) {
+            throw new \Exception('...');
```
Comment (bocharsky-bw) [inline src/AssetMapper/SassPublicPathAssetPathResolver.php:34]: I think it still would be great to have a dev-friendly message here to avoid confusing

**Comment by WebMamba** [comment]
@bocharsky-bw I do not have the right to merge this PR, and I think this is ready to go 😁

**Comment by bocharsky-bw** [comment]
Sorry for the delay, I hoped for @weaverryan final review on it.

Thanks!

File: config/services.php [hunk 3]
```diff
@@ -35,10 +37,18 @@
...
                 service('sass.builder'),
             ])
+
         ->set('sass.public_asset_path_resolver', SassPublicPathAssetPathResolver::class)
         ->decorate('asset_mapper.public_assets_path_resolver')
         ->args([
             service('.inner')
         ])
+
+        ->set('sass.listener.pre_assets_compile', PreAssetsCompileEventListener::class)
```
Comment (ogizanagi) [inline config/services.php:47]: The argument is missing: 

```
            ->args([
                service('sass.builder')
            ])
```

Comment (bocharsky-bw) [reply config/services.php:47]: Sounds valid. Would you like to create a PR?

//cc @WebMamba

**Pull request merged**